rpcs3/Emu/NP/clans_client.cpp

```diff
@@ -0,0 +1,1070 @@
+	{
+		if (idm::remove<clan_request_ctx>(reqId))
+			return SceNpClansError::SCE_NP_CLANS_SUCCESS;
+
+		return SceNpClansError::SCE_NP_CLANS_ERROR_INVALID_ARGUMENT;
+	}
+
+	SceNpClansError clans_client::sendRequest(u32 reqId, ClanRequestAction action, ClanManagerOperationType opType, pugi::xml_document* xmlBody, pugi::xml_document* outResponse)
+	{
+		auto ctx = idm::get_unlocked<clan_request_ctx>(reqId);
+
+		if (!ctx || !ctx->curl)
+			return SCE_NP_CLANS_ERROR_NOT_INITIALIZED;
+
+		CURL* curl = ctx->curl;
+
+		ClanRequestType reqType = ClanRequestType::FUNC;
+		pugi::xml_node clan = xmlBody->child("clan");
+		if (clan && clan.child("ticket"))
+		{
+			reqType = ClanRequestType::SEC;
+		}
+
+		std::string host = g_cfg_clans.get_host();
+		std::string protocol = g_cfg_clans.get_use_https() ? "https" : "http";
+		std::string url = fmt::format("%s://%s/clan_manager_%s/%s/%s", protocol, host, opType, reqType, action);
+
+		std::ostringstream oss;
+		xmlBody->save(oss, "\t", 8U);
+
+		std::string xml = oss.str();
+
+		char err_buf[CURL_ERROR_SIZE];
+		err_buf[0] = '\0';
+
+		std::vector<char> response_buffer;
+
+		curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, curlWriteCallback);
+		curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_buffer);
+		curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, err_buf);
+
+		curl_easy_setopt(curl, CURLOPT_POST, 1);
+		curl_easy_setopt(curl, CURLOPT_POSTFIELDS, xml.c_str());
+		curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, xml.size());
+
+		CURLcode res = curl_easy_perform(curl);
+
+		if (res != CURLE_OK)
+		{
+			outResponse = nullptr;
+			clan_log.error("curl_easy_perform() failed: %s", curl_easy_strerror(res));
+			clan_log.error("Error buffer: %s", err_buf);
+			return SCE_NP_CLANS_ERROR_BAD_REQUEST;
+		}
+
+		response_buffer.push_back('\0');
+
+		pugi::xml_parse_result xml_res = outResponse->load_string(response_buffer.data());
+		if (!xml_res)
+		{
+			clan_log.error("XML parsing failed: %s", xml_res.description());
+			return SCE_NP_CLANS_ERROR_BAD_RESPONSE;
+		}
+
+		pugi::xml_node clanResult = outResponse->child("clan");
+		if (!clanResult)
+			return SCE_NP_CLANS_ERROR_BAD_RESPONSE;
+
+		pugi::xml_attribute result = clanResult.attribute("result");
+		if (!result)
+			return SCE_NP_CLANS_ERROR_BAD_RESPONSE;
+
+		std::string result_str = result.as_string();
+		if (result_str != "00")
+			return static_cast<SceNpClansError>(0x80022800 | std::stoul(result_str, nullptr, 16));
+
+		return SCE_NP_CLANS_SUCCESS;
+	}
+
+	std::string clans_client::getClanTicket(np::np_handler& nph)
+	{
+		const auto& npid = nph.get_npid();
+
+		const char* service_id = CLANS_SERVICE_ID;
+		const unsigned char* cookie = nullptr;
+		const u32 cookie_size = 0;
+		const char* entitlement_id = CLANS_ENTITLEMENT_ID;
+		const u32 consumed_count = 0;
+
+		nph.req_ticket(0x00020001, &npid, service_id, cookie, cookie_size, entitlement_id, consumed_count);
```

Comment:
Is there a reason for requesting the ticket each time? Shouldn't you request the ticket once and reuse it each time?